mitiq/interface/conversions.py

```diff
@@ -296,68 +297,83 @@ def new_scaling_function(
-            _remove_identity_from_idle(scaled_circuit, idle_qubits)
-            if circuit.cregs and not scaled_circuit.cregs:
-                scaled_circuit.add_register(*circuit.cregs)
-
-            for q, c in _measurement_order(circuit):
-                scaled_circuit.measure(q, c)
-
-        return scaled_circuit
-
-    return new_scaling_function
+            out_circuits = [out_circuit]
+
+        circuits_to_return = []
+        for out_circuit in out_circuits:
+            # Post atomic conversion
+            # PyQuil: Restore declarations, measurements, and metadata.
+            if "pyquil" in out_circuit.__module__:
+                from pyquil import Program
+                from pyquil.quilbase import Declare, Measurement
+
+                circuit = cast(Program, circuit)
+
+                # Grab all measurements from the input circuit.
+                measurements = [
+                    instr
+                    for instr in circuit.instructions
+                    if isinstance(instr, Measurement)
+                ]
+
+                # Remove memory declarations added from Cirq-pyQuil conversion.
+                new_declarations = {
+                    k: v
+                    for k, v in out_circuit.declarations.items()
+                    if k == "ro" or v.memory_type != "BIT"
+                }
+                new_declarations.update(circuit.declarations)
+
+                # Delete all declarations and measurements.
+                instructions = [
+                    instr
+                    for instr in out_circuit.instructions
+                    if not (isinstance(instr, (Declare, Measurement)))
+                ]
+
+                # Add back original declarations and measurements.
+                out_circuit = Program(
+                    list(new_declarations.values())
+                    + instructions
+                    + measurements
+                )
+
+                # Set the number of shots to the input circuit.
+                out_circuit.num_shots = circuit.num_shots
+
+            # Qiskit: Keep the same register structure and measurement order.
+            if "qiskit" in out_circuit.__module__:
+                from mitiq.interface.mitiq_qiskit.conversions import (
+                    _measurement_order,
+                    _remove_identity_from_idle,
+                    _transform_registers,
+                )
+
+                out_circuit.remove_final_measurements()
+                _transform_registers(out_circuit, new_qregs=circuit.qregs)
+                _remove_identity_from_idle(out_circuit, idle_qubits)
+                if circuit.cregs and not out_circuit.cregs:
+                    out_circuit.add_register(*circuit.cregs)
+
+                for q, c in _measurement_order(circuit):
+                    out_circuit.measure(q, c)
+            circuits_to_return.append(out_circuit)
+
+        if not one_to_many:
+            assert len(circuits_to_return) == 1
```

Comment:
I don't think we normally have `assert` statements in our non-test code. I'm not opposed to it, but wanted to flag this in case it was left over from testing.

Reply:
Once I tried reading something online about `assert` vs `exceptions`. 
If I understand well, `assert` can be in the code to "assert" things which, unless we are missing something, must be always true. In practice, there shouldn't be a simple way of triggering that error. But since you are not 100% sure, you add the assert statement as a double check.

Exceptions instead are designed for "triggerable" errors caused by using a function in the wrong way, typically by a user of the code. So the error condition can be both true or false, depending on good or bad usage of the function.

Again, I am not sure if this makes sense. But it sounds reasonable.

Reply:
Ah okay that's a good distinction to make. Thanks for flushing that out. Happy to keep it (and this practice moving forward).